### PR TITLE
fork-tree: fix tree rebalancing

### PR DESCRIPTION
**Pull request by andresilva**
We want to keep the fork tree nodes sorted by subtree depth, this is an heuristic since the deepest subtree will most likely translate to the longest chain, in which case any query on the tree will most likely want to target the longest branch. To this end, after inserting a new node in the tree we make sure it is rebalanced. Currently we were only triggering the rebalancing when the new node being inserted was a new root, and not when it was a leaf node part of an already existing subtree.

I also fixed the initial state of the tree iterator which would mess up the ordering.

### CODE REVIEW

**Review by octol — APPROVED** [review]

File: utils/fork-tree/src/lib.rs [hunk 1]
```diff
@@ -939,6 +944,10 @@ mod test {
 		//   — J - K
 		//
 		// (where N is not a part of fork tree)
+		//
+		// NOTE: the tree will get automatically rebalance on import and won't be laid out like the
+		// diagram above. the children will be ordered by subtree depth and the longest branches
+		// will be on the leftmost side of the tree.
```
Comment (cheme) [inline utils/fork-tree/src/lib.rs:950]: I was thinking a post rebalance diagram would be good, but the explanation is enough (I tried to update the schema but got lost with my bad ascii writing whent on the other hand the iter test was easy to read from the old schema and this comment).

**Review by cheme — APPROVED** [review]
LGTM

File: utils/fork-tree/src/lib.rs (hunk 1 above)
Comment (andresilva) [reply utils/fork-tree/src/lib.rs:950]: Yeah I was thinking of doing the same thing but then opted to just add the comment instead :P

**Comment by andresilva** [comment]
bot merge

**Pull request merged**